### PR TITLE
feat/python: Update NodeInfo

### LINKED ISSUE #1017: Update InfoResponse - python
https://github.com/iotaledger/tips-draft/pull/24

### LINKED ISSUE #1135: Adapt ProtocolParameters - Python
https://github.com/iotaledger/tips-draft/pull/25

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

Updates NodeInfo and subtypes

## Links to any relevant issues

fixes #1017
fixes #1135

### CODE REVIEW

File: bindings/python/iota_sdk/types/node_info.py [hunk 1]
```diff
@@ -7,36 +7,37 @@
...
-    pruning_index: int
+    accepted_tangle_time: str
+    relative_accepted_tangle_time: int
+    confirmed_tangle_time: int
+    relative_confirmed_tangle_time: int
+    latest_commitment_id: HexStr
+    # TODO Replace with a proper SlotIndex type
+    latest_finalized_slot: str
+    # TODO Replace with a proper SlotIndex type
+    latest_accepted_block_slot: str
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/node_info.py:36]: some of these fields are optional

File: bindings/python/iota_sdk/types/node_info.py [hunk 2]
```diff
@@ -45,84 +46,198 @@ class NodeInfoMetrics:
...
+        mana_decay_factor_epochs_sum_exponent: The scaling of mana_decay_factor_epochs_sum expressed as an exponent of 2.
+        staking_unbonding_period: The unbonding period in epochs before an account can stop staking.
+        liveness_threshold: Determine if a block is eligible by evaluating issuing_time and commitments in its past cone to ATT and last_committed_slot respectively.
+        min_committable_age: Min_committable_age is the minimum age relative to the accepted tangle time slot index that a slot can be committed.
+        max_committable_age: Max_committable_age is the maximum age for a slot commitment to be included in a block relative to the slot index of the block issuing time.
+        epoch_nearing_threshold: Determine the slot that should trigger a new committee selection for the next and upcoming epoch.
+        congestion_control_parameters: Congestion Control Parameters defines the parameters used to calculate the Reference Mana Cost (RMC).
+        version_signaling: The version signaling parameters.
+    """
+    version: int
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/node_info.py:207]: Missing type field?

File: bindings/python/iota_sdk/types/node_info.py [hunk 3]
```diff
@@ -45,84 +46,198 @@ class NodeInfoMetrics:
...
-    token_supply: str
-    version: int
     rent_structure: RentStructure
+    work_score_structure: WorkScoreStructure
+    token_supply: str
+    genesis_unix_timestamp: int
+    slot_duration_in_seconds: int
+    slots_per_epoch_exponent: int
+    mana_bits_count: int
+    mana_generation_rate: int
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/node_info.py]: Mana structure?

File: bindings/python/iota_sdk/types/node_info.py [hunk 4]
```diff
@@ -45,84 +46,198 @@ class NodeInfoMetrics:
...
+    mana_generation_rate_exponent: int
+    mana_decay_factors: List[int]
+    mana_decay_factors_exponent: int
+    mana_decay_factor_epochs_sum: int
+    mana_decay_factor_epochs_sum_exponent: int
+    staking_unbonding_period: str
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/node_info.py:220]: missing validation blocks per slot?

File: bindings/python/iota_sdk/types/node_info.py [hunk 5]
```diff
@@ -45,84 +46,198 @@ class NodeInfoMetrics:
...
+    scheduler_rate: int
+    min_mana: str
+    max_buffer_size: int
 
 
 @json
 @dataclass
-class NodeInfoProtocol:
-    """Protocol info.
+class VersionSignaling:
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/node_info.py:141]: VersionSignalingParameters

File: bindings/python/iota_sdk/types/node_info.py (hunk 1 above)
Comment (kwek20) [reply bindings/python/iota_sdk/types/node_info.py:36]: Not according to here: https://github.com/iotaledger/tips-draft/pull/24/files#diff-04aff9a5004dd8023df5735c866f69254359635764a00e9d9215500a9676319bR2575

But i think im at the wrong TIP :/

**Review by DaughterOfMars — COMMENTED** [review]
Probably should add a test that does the same thing as the test in protocol.rs

File: bindings/python/iota_sdk/types/node_info.py (hunk 5 above)
Comment (kwek20) [reply bindings/python/iota_sdk/types/node_info.py:141]: Not according to tip

**Comment by kwek20** [comment]
> Probably should add a test that does the same thing as the test in protocol.rs

I dont think thats beneficial, seems to just make a random protocol with random fields in rust.
We could use a nodeinfo test i guess, but out of scope

**Comment by DaughterOfMars** [comment]
It's to verify the structure against the TIP data...

**Comment by kwek20** [comment]
> It's to verify the structure against the TIP data...

Yea i understand, but as we pass rust ProtocolParams, we get that for free right?

File: bindings/python/iota_sdk/types/node_info.py [hunk 6]
```diff
@@ -45,84 +46,218 @@ class NodeInfoMetrics:
...
-    referenced_rate: float
+    confirmed_blocks_per_second: float
+    confirmation_rate: float
 
 
 @json
 @dataclass
 class RentStructure:
-    """Rent structure for the storage deposit.
+    """The rent structure according to TIP-19.
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/node_info.py]: Don't reference TIP19, it's stardust

Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/node_info.py]: Comment before was nice, just revert

File: bindings/python/iota_sdk/types/node_info.py [hunk 7]
```diff
@@ -60,69 +61,203 @@ class RentStructure:
...
+    """The protocol parameters.
+
+    Attributes:
+        type: Set to value 0 to denote a IOTA 2.0 protocol parameter.
+        version: Protocol version used by the network.
+        network_name: The Name of the network from which the networkId is derived.
+        bech32_hrp: Tells whether the node supports mainnet or testnet addresses.
+                    Value `iota` indicates that the node supports mainnet addresses.
+                    Value `atoi` indicates that the node supports testnet addresses.
+        rent_structure: The rent structure according to TIP-19.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/node_info.py]: Also wrong TIP

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/python/iota_sdk/types/node_info.py [hunk 8]
```diff
@@ -7,36 +7,37 @@
...
+        latest_accepted_block_slot: The slot index of the latest accepted block.
+        latest_confirmed_block_slot: The slot index of the latest confirmed block.
+        pruning_epoch: The index of the slot before which the tangle history is pruned.
     """
     is_healthy: bool
-    latest_milestone: NodeInfoMilestone
-    confirmed_milestone: NodeInfoMilestone
-    pruning_index: int
+    accepted_tangle_time: str
+    relative_accepted_tangle_time: int
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/node_info.py]: Why int? same for following fields

Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/node_info.py]: Or, why is the previous field str?

File: bindings/python/iota_sdk/types/node_info.py [hunk 9]
```diff
@@ -60,69 +61,203 @@ class RentStructure:
...
+    type: int
+    version: int
     network_name: str
     bech32_hrp: str
-    token_supply: str
-    version: int
     rent_structure: RentStructure
+    work_score_structure: WorkScoreStructure
+    token_supply: str
+    genesis_unix_timestamp: int
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/node_info.py]: Int ?

Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/node_info.py]: Can you review all int/string ? It doesn't looks super correct

**Comment by DaughterOfMars** [comment]
> > It's to verify the structure against the TIP data...
> 
> Yea i understand, but as we pass rust ProtocolParams, we get that for free right?

No, this is an independent structure. We have no tests to validate it's structure.

**Review by thibault-martinez — APPROVED** [review]

**Comment by thibault-martinez** [comment]
> > > It's to verify the structure against the TIP data...
> > 
> > 
> > Yea i understand, but as we pass rust ProtocolParams, we get that for free right?
> 
> No, this is an independent structure. We have no tests to validate it's structure.

I agree

**Comment by kwek20** [comment]
> > > It's to verify the structure against the TIP data...
> > 
> > 
> > Yea i understand, but as we pass rust ProtocolParams, we get that for free right?
> 
> No, this is an independent structure. We have no tests to validate it's structure.

Made an issue https://github.com/iotaledger/iota-sdk/issues/1338

**Review by Thoralf-M — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**